static-checks/ansible/allowed-modules/test.py

```diff
@@ -0,0 +1,121 @@
+#!/usr/bin/python3
+
+import os
+import re
+import tempfile
+import yaml
+import contextlib
+from pathlib import Path
+
+from lib import util, results
+
+
+# Obtained from
+# https://docs.ansible.com/ansible/latest/reference_appendices/playbooks_keywords.html#task
+ansible_reserved_keywords = {
+    'action', 'any_errors_fatal', 'args', 'async', 'become', 'become_exe', 'become_flags',
+    'become_method', 'become_user', 'changed_when', 'check_mode', 'collections', 'connection',
+    'debugger', 'delay', 'delegate_facts', 'delegate_to', 'diff', 'environment', 'failed_when',
+    'ignore_errors', 'ignore_unreachable', 'local_action', 'loop', 'loop_control',
+    'module_defaults', 'name', 'no_log', 'notify', 'poll', 'port', 'register', 'remote_user',
+    'retries', 'run_once', 'tags', 'throttle', 'timeout', 'until', 'vars', 'when', 'with_items'
+}
+
+
+@contextlib.contextmanager
+def select_all_rules(datastream):
+    """
+    Given an XML file path as 'datastream' enable all the rules in the datastream.
+
+    Use as a context manager, returning a filesystem path to the
+    modified datastream.
+    """
+    exprs = set()
+    exprs.add(re.compile(r'<.*Rule.*id="'))
+    exprs.add(re.compile(r'<.*select.*idref="'))
+
+    with open(datastream) as ds:
+        with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as new_ds:
+            for line in ds:
+                if any(x.search(line) for x in exprs):
+                    line = line.replace('selected="false"', 'selected="true"')
+                new_ds.write(line)
+            new_ds.flush()
+            util.log(f"Saving modified {datastream} with all rules selected as {new_ds.name}")
+            yield Path(new_ds.name)
+
+
+def process_task(task, all_allowed_modules):
+    original_keywords = set(task.keys())
+    found_allowed_modules = set()
+
+    if 'block' in original_keywords:
+        for block_task in task['block']:
+            found_allowed_modules.update(process_task(block_task, all_allowed_modules))
+    else:
+        keywords = set()
+        for keyword in original_keywords:
+            keywords.add(keyword.replace('ansible.builtin.', ''))
+        allowed_module = keywords.intersection(all_allowed_modules)
+        if allowed_module:
+            found_allowed_modules.update(allowed_module)
+        else:
+            # No allowed module found in 'keywords' means there is a module
+            # which is forbidden and we obtain it by filtering out all the
+            # reserved keywords for Ansible tasks which will leave us with
+            # one item and that should be the name of that forbidden module.
+            forbidden_module = keywords.difference(ansible_reserved_keywords).pop()
+            task_name = task['name'] if 'name' in task else ''
+            results.report('fail', forbidden_module, f'forbidden module in task \'{task_name}\'')
```

Comment:
I would avoid reporting kind of useless `note` for the results (here and for `pass` as well). A note being present in the HTML distorts the output somewhat and is just visual clutter in this case.

The test is named `allowed-modules`, so any failures should be forbidden modules, and any passes allowed modules, no need to verbosely state that in the note.

But I'll leave the decision up to you.

Reply:
Ok, I will update it.